### ISSUE TITLE
Quark function generators for user-defined quarks?

### ISSUE THREAD

**Issue by muzimuzhi** [latex3/latex3#1621]
Built-in quarks have test and recursion functions to work with, but for new quarks defined by `\quark_new:N`, there are no such functions. Providing/Exposing quark function generators can make  user-defined quarks more useable.

For example, `\q_no_value` has `\quark_if_no_value_p:N(TF)` test function accompanied, and `\q_recursion_tail` has `\quark_if_recursion_tail_stop:N` recursion function. But after `\quark_new:N \q__mymod`, currently the new quark can only used in
```tex
\tl_if_eq:NNTF <token> \q__mymod {<true>} {<false>}
```

The `\__kernel_quark_new_test:N` and `\__kernel_quark_new_conditional:Nn` defined in `l3quark.dtx` are already used by other l3kernel modules to define functions for their internal quarks. Maybe these two kernel functions can be exposed.

**Comment by josephwright** [comment]
I've wondered much the same: let's see what others feel

**Comment by FrankMittelbach** [comment]
no problem with that I think

**Comment by josephwright** [comment]
I'm thinking `\quark_new_conditional:Nn` and `\quark_new_recursion_conditionals:N` - does that sound OK?

**Comment by FrankMittelbach** [comment]
why plural and singular?

**Comment by josephwright** [comment]
> why plural and singular?

`\quark_new_conditional:Nn` generates `\<module>_quark_if_<name>:n(TF)` or  `\<module>_quark_if_<name>:N(TF)`, depending on the first argument. 

`\quark_new_recursion_conditionals:N` generates one of six possibles, at least at present. I wonder if it would be better to generate all six - see next comment.

**Comment by josephwright** [comment]
I suspect the reason we kept these internal is they need a rather precise form of the base name. We might want to broaden that out or re-work the interface a little - I'll see what makes sense. (I had though I could avoid a PR but looking again there's some stuff to decide so I will go that way.)

**Comment by FrankMittelbach** [comment]
> > why plural and singular?
> 
> `\quark_new_conditional:Nn` generates `\<module>_quark_if_<name>:n(TF)` or `\<module>_quark_if_<name>:N(TF)`, depending on the first argument.
> 
> `\quark_new_recursion_conditionals:N` generates one of six possibles, at least at present. I wonder if it would be better to generate all six - see next comment.

if both generate one then it should be singular shouldn't it? that was my point.

**Comment by josephwright** [comment]
> > > why plural and singular?
> > 
> > 
> > `\quark_new_conditional:Nn` generates `\<module>_quark_if_<name>:n(TF)` or `\<module>_quark_if_<name>:N(TF)`, depending on the first argument.
> > `\quark_new_recursion_conditionals:N` generates one of six possibles, at least at present. I wonder if it would be better to generate all six - see next comment.
> 
> if both generate one then it should be singular shouldn't it? that was my point.

Yeah, I'd missed something there - let me firm up a proposal later today

**Comment by josephwright** [comment]
OK, I've been looking over this and I see why we kept them internal - the interface is a bit odd. The current `\__kernel_quark_new_conditional:Nn` is reminiscent of `\prg_new_conditional:Nnn` or similar, in that the `N`-type argument could arguably be (multiple)  `n`-type:
```latex
\__kernel_quark_new_conditional:Nn \mymodule_quark_if_nil:N { T , F , TF }
```
could be 
```latex
\quark_new_conditional:nnNn { mymodule } { nil } N { T , F , TF }
```
or
```latex
\quark_new_conditional:NNn \q_mymodule_nil N { T , F , TF }
```
or even
```latex
\quark_new_conditional:Nn \q_mymodule_nil { T , F , TF }
```
(and generate both `n`- and `N`-type conditionals).

The situation with `\__kernel_quark_new_test:N` is more complicated as it can generate six different outcomes (!).
```latex
\__kernel_quark_new_test:N \mymodule_if_recursion_tail_stop:N`
```
generates `\quark_if_recursion_tail_stop:N`, so here the token name is there but it has to contain the `if_recursion_tail` part and have `stop`/`stop_do`/`break`, etc. match the arg spec. So one might argue that this is basically OK and just rename (but it's `test` not `conditional` as we don't have `TF`), or for
```latex
\quark_new_recursion_test:nn { mymodule } { stop }
```
or perhaps again something that does 'all the work'
```latex
\quark_recursion_new:n { mymodule }
```
to auto-generate the `tail` and `stop` quarks and all six possible tests in one go.
for

**Comment by josephwright** [comment]
Or
```latex
\quark_recursion_new:N \q_mymodule_tail
\quark_recursion_new:NN \q_mymodule_tail \q_mymodule_stop
```
i.e. including one or (perhaps) both quarks explicitly

**Comment by josephwright** [comment]
My favoured would be 
```latex
\quark_new_conditional:Nn \q_mymodule_nil { T , F , TF }
\quark_recursion_new:NN \q_mymodule_tail \q_mymodule_stop
```
and auto-generate, BTW (actually, I'd drop the `n`-type argument in the first one, cf. https://github.com/latex3/latex3/pull/1619, but that's an ongoing discussion).
